### PR TITLE
Encode spans relative to the enclosing item -- enable on nightly

### PR DESCRIPTION
**Pull request by cjgillot**
Follow-up to #84373 with the flag `-Zincremental-relative-spans` set by default.

This PR seeks to remove one of the main shortcomings of incremental: the handling of spans.
Changing the contents of a function may require redoing part of the compilation process for another function in another file because of span information is changed.
Within one file: all the spans in HIR change, so typechecking had to be re-done.
Between files: spans of associated types/consts/functions change, so type-based resolution needs to be re-done (hygiene information is stored in the span).

The flag `-Zincremental-relative-spans` encodes local spans relative to the span of an item, stored inside the `source_span` query.

Trap: stashed diagnostics are referenced by the "raw" span, so stealing them requires to remove the span's parent.

In order to avoid too much traffic in the span interner, span encoding uses the `ctxt_or_tag` field to encode:
- the parent when the `SyntaxContext` is 0;
- the `SyntaxContext` when the parent is `None`.
Even with this, the PR creates a lot of traffic to the Span interner, when a Span has both a LocalDefId parent and a non-root SyntaxContext. They appear in lowering, when we add a parent to all spans, including those which come from macros, and during inlining when we mark inlined spans.

The last commit changes how queries of `LocalDefId` manage their cache. I can put this in a separate PR if required.

Possible future directions:
- validate that all spans are marked in HIR validation;
- mark macro-expanded spans relative to the def-site and not the use-site.

### CODE REVIEW

**Comment by petrochenkov** [comment]
I'm not sure what is the status here.
@rustbot author

**Comment by Dylan-DPC** [comment]
Closing this as it is inactive and bitrotted

**Comment by cjgillot** [comment]
I do not have any more ideas to improve on this. 
The regressions are due to increased trafic with the Span interner when a span has both a LocalDefId parent and a non-root SyntaxContext. 
They appear in lowering, when we add a parent to all spans, and during inlining when we mark inlined spans.

**Comment by petrochenkov** [comment]
>The last commit changes how queries of `LocalDefId` manage their cache. I can put this in a separate PR if required.

Yeah, I think it'd be better to move the query system commits (https://github.com/rust-lang/rust/pull/84762/commits/4d1e0f7541481ccbd821f86a01d277724523168b and https://github.com/rust-lang/rust/pull/84762/commits/9f4e9be97c785d8fdf2f010cc42331cbc5960ab3) to a separate PR, with someone else as a reviewer.

File: src/test/incremental/thinlto/cgu_keeps_identical_fn.rs [hunk 1]
```diff
@@ -3,29 +3,20 @@
 // ends up with any spans in its LLVM bitecode, so LLVM is able to skip
 // re-building any modules which import 'inlined_fn'
 
-// revisions: cfail1 cfail2 cfail3 cfail4 cfail5 cfail6
-// [cfail4]compile-flags: -Zincremental-relative-spans
-// [cfail5]compile-flags: -Zincremental-relative-spans
-// [cfail6]compile-flags: -Zincremental-relative-spans
+// revisions: cfail1 cfail2 cfail3
```
Comment (petrochenkov) [inline src/test/incremental/thinlto/cgu_keeps_identical_fn.rs:6]: The test diffs are unreadable.
Perhaps it's better to remove non-default revisions in one commit, then flip the switch, and update the tests for the new default behavior in another commit.

Comment (cjgillot) [reply src/test/incremental/thinlto/cgu_keeps_identical_fn.rs:6]: I think that removing the `-Zincremental-ignore-spans` option is a bad idea. Most of the test fallout came from there. The commits are much simpler now.

**Comment by cjgillot** [comment]
Cache changes have been submitted as https://github.com/rust-lang/rust/pull/103808.

**Comment by petrochenkov** [comment]
I'll block this on https://github.com/rust-lang/rust/pull/103808, then do a final per run after it's unblocked, then probably start a compiler team FCP.

**Comment by petrochenkov** [comment]
Ah, I also had an idea, not sure whether it was tried before or not.

Right now a span with a parent is represented as `(parent, lo, len, ctxt)`.

What if we represent it as `(parent, Δlo, len, Δctxt)` instead, where delta is difference from the base parent's span?
These values would have a very different distribution from `(parent, lo, len, ctxt)`, `Δlo` and `Δctxt` in particular could be expected to be small. So we can pack them into the `Span`'s 64 bits more effectively.

The drawback is that you'd need a `tcx` to get an absolute span from the "delta" span.

**Comment by cjgillot** [comment]
@petrochenkov thanks for the idea.  I've been trying to implement this, but I think it would be preferable to separate this change to a follow-up PR to benchmark it on its own.

**Comment by petrochenkov** [comment]
Ok, let's start an FCP.
This PR trades small perf regressions in all cases for large perf benefits in incremental cases.
Benchmarking results - https://github.com/rust-lang/rust/pull/84762#issuecomment-1326957120.
More detailed description - https://github.com/rust-lang/rust/pull/84762#issue-873086666.

@rfcbot fcp merge

**Comment by michaelwoerister** [comment]
Regarding performance, I think this should be enabled.

But I'm a bit worried about the potential for subtle bugs here. Spans + incremental just has a track record of introducing those. Is there a way to get more real-world testing for the feature before it's enabled by default? Maybe enable it by default just on nightly for a couple of cycles?

**Comment by cjgillot** [comment]
@michaelwoerister I agree. I can enable this on nightly only, and then by default in 1.70.0.

**Comment by michaelwoerister** [comment]
Sounds good, @cjgillot 🙂

**Comment by apiraino** [comment]
@petrochenkov I had removed `S-waiting-on-team` on purpose last week after [T-compiler meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-12-08/near/314697466). Makes sense?

**Comment by petrochenkov** [comment]
@apiraino 
>Makes sense?

Not at all.
There should always be some "waiting on" status label, and in this case waiting-on-team is the most appropriate label because the FCP week is still a time for potential comments from the team.

**Comment by pnkfelix** [comment]
> @apiraino
> 
> > Makes sense?
> 
> Not at all. There should always be some "waiting on" status label, and in this case waiting-on-team is the most appropriate label because the FCP week is still a time for potential comments from the team.

I agree there should always be some "waiting on" status label.

In this case, it would have been reasonable to apply S-waiting-on-fcp (rather than S-waiting-on-team), since that is specifically for PRs that have entered FCP and are going through the ten day final comment period.

(And also because all but two of the team had indeed signed off already, so S-waiting-on-fcp is more accurate reflection of what we are waiting on, in my opinion.)

**Comment by petrochenkov** [comment]
>In this case, it would have been reasonable to apply S-waiting-on-fcp

Ah, okay, I haven't seen this label before.

**Comment by rylev** [comment]
There was an [FCP](https://github.com/rust-lang/rust/pull/84762#issuecomment-1327518372) on whether the performance regression in some benchmarks were worth it which passed.

@rustbot label perf-regression-triaged

**Comment by dtolnay** [comment]
This had what seems to be an unintended effect on JSON diagnostics: #106571.

**Pull request merged**